src/mlpack/methods/reinforcement_learning/replay/sumtree.hpp

```diff
@@ -0,0 +1,194 @@
+/**
+ * @file sumtree.hpp
+ * @author Xiaohong
+ *
+ * This file is an implementation of sumtree.
+ *
+ * reference:
+ * [1] https://github.com/openai/baselines/blob/master/baselines/common/segment_tree.py
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_SUMTREE_HPP
+#define MLPACK_METHODS_RL_SUMTREE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of SumTree.
+ *
+ * Build a Segment Tree like data structure.
+ * https://en.wikipedia.org/wiki/Segment_tree
+ *
+ * Used to maintain prefix-sum of an array.
+ *
+ */
+
+template<typename T>
+class SumTree
+{
+ public:
+  /**
+   * Default constructor.
+   */
+  SumTree():
+      capacity(0)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of SumTree class.
+   *
+   * @param capacity Size of data.
+   */
+  SumTree(size_t capacity):
+      capacity(capacity)
+  {
+    element = std::vector<T> (2 * capacity);
+  }
+
+  /**
+   * Set the data array with idx.
+   *
+   * @param idx The array idx to be changed.
+   * @param value The data that array with idx to be.
+   */
+  void Set(size_t idx, T value)
```

Comment:
We might want to use ``const`` if possible that applies to the other methods as well.